src/ape/pytest/fixtures.py

```diff
@@ -20,3 +24,78 @@ def chain(self) -> ChainManager:
     @pytest.fixture(scope="session")
     def project(self) -> ProjectManager:
         return self.project_manager
+
+    @pytest.fixture(scope="session")
+    def _session_isolation(self) -> Iterator[None]:
+        snapshot_id = None
+        try:
+            snapshot_id = self.chain_manager.snapshot()
+        except NotImplementedError:
+            self._warn_for_unimplemented_snapshot()
+
+        yield
+
+        if snapshot_id is not None and snapshot_id in self.chain_manager._snapshots:
+            self.chain_manager.restore(snapshot_id)
+
+    @pytest.fixture(scope="package")
+    def _package_isolation(self, _session_isolation) -> Iterator[None]:
+        snapshot_id = None
+        try:
+            snapshot_id = self.chain_manager.snapshot()
+        except NotImplementedError:
+            self._warn_for_unimplemented_snapshot()
+
+        yield
+
+        if snapshot_id is not None and snapshot_id in self.chain_manager._snapshots:
+            self.chain_manager.restore(snapshot_id)
+
+    @pytest.fixture(scope="module")
+    def _module_isolation(self, _package_isolation) -> Iterator[None]:
+        snapshot_id = None
+        try:
+            snapshot_id = self.chain_manager.snapshot()
+        except NotImplementedError:
+            self._warn_for_unimplemented_snapshot()
+
+        yield
+
+        if snapshot_id is not None and snapshot_id in self.chain_manager._snapshots:
+            self.chain_manager.restore(snapshot_id)
+
+    @pytest.fixture(scope="class")
+    def _class_isolation(self, _module_isolation) -> Iterator[None]:
+        snapshot_id = None
```

Comment:
Could this logic be abstracted and shared using an internal helper
method _snapshot() -> SnapShotID that handles the warning and try/catch?

Reply:
i think so :thinking: if anything I can just not use `pytest.fixture` as a decorator, I'll give it a try though (too much bloat :laughing: )